File: web/src/lib/fleet.ts
// One socket per machine, carrying what the client used to fetch on a timer.
//
// The poll loop asked every machine for its session list every few seconds. That
// was wrong twice over: too slow, because the sidebar reports what each agent is
// doing and a status board that lags by up to eight seconds will tell you an
// agent is working after it stopped to ask you something; and too much, because
// the answer usually had not changed and the cost was a request per machine per
// tick, growing with the size of the fleet rather than with how much was going on.
//
// This does not replace the REST endpoints. They remain the way the app loads and
// the way it recovers: the socket is an accelerator, and everything still works
// with it disconnected, just at the old cadence. That is deliberate, because a
// machine can be reachable for HTTP and not for WebSocket (a proxy, an old
// server without the route) and the fleet must not go dark when it is.

import type { Meta, Stats } from './types'

// Mirrors fleetMsg in internal/server/fleet.go; keep the two in step by hand, as
// with the rest of the wire contract.
export interface FleetMsg {
  t: 'sessions' | 'stats'
  sessions?: Meta[]
  stats?: Stats
}

export interface FleetHandlers {
  onSessions: (sessions: Meta[]) => void
  onStats: (stats: Stats) => void
  // onOpen and onClose let the caller keep polling only while the push is down,
  // which is what makes this an accelerator rather than a second source of truth.
  onOpen?: () => void
  onClose?: () => void
}

// Reconnect backoff. Starts fast because the common disconnect is a laptop lid or
// a phone unlocking, where the machine is already there; grows so an unreachable
// peer is not hammered.
const RETRY_MIN = 1_000
const RETRY_MAX = 30_000

export class FleetSocket {
  private ws: WebSocket | null = null
  private retry = RETRY_MIN
  private timer: ReturnType<typeof setTimeout> | undefined
  private stopped = false
  private readonly url: string

  constructor(
    base: string,
    private handlers: FleetHandlers,
  ) {
    // base is '' for the hub (this origin) and a full origin for a peer, the same
    // convention every call in api.ts uses.
    const origin = base || location.origin
    this.url = origin.replace(/^http/, 'ws') + '/ws/fleet'
  }

  start() {
    this.stopped = false
    this.open()
  }

  stop() {
    this.stopped = true
    clearTimeout(this.timer)
    // Close the socket without waiting for a reply: this runs when a machine is
    // removed or the app tears down, and there is nothing to say.
    this.ws?.close()
    this.ws = null
  }

  get connected(): boolean {
    return this.ws?.readyState === WebSocket.OPEN
  }

  private open() {
    if (this.stopped) return
    let ws: WebSocket
    try {
      ws = new WebSocket(this.url)
    } catch {
      // A malformed origin, which is a permanent failure rather than a blip: back
      // off to the maximum instead of retrying in a second forever.
      this.retry = RETRY_MAX
      this.schedule()
      return
    }
    this.ws = ws

    ws.onopen = () => {
      this.retry = RETRY_MIN
      this.handlers.onOpen?.()
    }
    ws.onmessage = (e) => {
      let msg: FleetMsg
      try {
        msg = JSON.parse(e.data as string)
      } catch {
        return
      }
      if (msg.t === 'sessions' && msg.sessions) this.handlers.onSessions(msg.sessions)
      else if (msg.t === 'stats' && msg.stats) this.handlers.onStats(msg.stats)
    }
    // One handler for both: onerror is always followed by onclose, and reconnecting
    // from each would open two sockets.
    ws.onclose = () => {
      if (this.ws !== ws) return // superseded by a newer socket
      this.ws = null
      this.handlers.onClose?.()
      this.schedule()
    }
    ws.onerror = () => {
      /* onclose follows; nothing to do here */
    }
  }

  private schedule() {
    if (this.stopped) return
    clearTimeout(this.timer)
    this.timer = setTimeout(() => this.open(), this.retry)
    this.retry = Math.min(this.retry * 2, RETRY_MAX)
  }
}

File: internal/server/fleet.go
package server

import (
	"context"
	"encoding/json"
	"net/http"
	"sync"
	"time"

	"github.com/coder/websocket"
	"github.com/hegade/kunai/internal/session"
)

// The fleet socket: one per machine, carrying the things a client used to ask for
// on a timer.
//
// The client polled every machine for its session list every few seconds. That is
// wrong in two directions at once. It is too slow, because the sidebar reports
// what each agent is doing and a status board that lags by up to eight seconds
// will tell you an agent is working after it has stopped to ask you something. And
// it is too much, because the answer usually has not changed, and the cost is a
// request per machine per tick, so it grows with the size of your fleet rather
// than with how much is happening.
//
// A push fixes both: nothing on the wire while nothing happens, and a change
// arrives in one round trip instead of on the next beat. The session list is sent
// whole rather than as a delta -- it is a handful of small objects, and a whole
// list cannot drift out of sync with the server the way a stream of patches can.
//
// Stats stay on a timer, because CPU and memory are sampled rather than evented,
// but the timer now lives on the server: it ticks only while somebody is
// connected, and one tick serves every client on that machine.
const (
	// How long to gather changes before sending. A turn ending fires several
	// changes in a few milliseconds (state to idle, the session's own events, a
	// reap), and each would otherwise be its own list send.
	fleetCoalesce = 120 * time.Millisecond
	// Stats cadence while a client is watching. The client used to fetch these
	// every ~16s per machine.
	fleetStatsEvery = 10 * time.Second
)

// fleetHub fans machine-level changes out to connected clients.
type fleetHub struct {
	mu   sync.Mutex
	subs map[chan struct{}]bool
}

func newFleetHub() *fleetHub {
	return &fleetHub{subs: make(map[chan struct{}]bool)}
}

// notify wakes every subscriber. The channel is depth-1 and the send is
// non-blocking, so a client that is slow to read coalesces into one wake-up
// instead of applying back-pressure to the session that changed.
func (h *fleetHub) notify() {
	h.mu.Lock()
	defer h.mu.Unlock()
	for ch := range h.subs {
		select {
		case ch <- struct{}{}:
		default:
		}
	}
}

func (h *fleetHub) subscribe() chan struct{} {
	ch := make(chan struct{}, 1)
	h.mu.Lock()
	h.subs[ch] = true
	h.mu.Unlock()
	// Seeded so a new subscriber sends its first snapshot immediately without a
	// special case in the loop.
	ch <- struct{}{}
	return ch
}

func (h *fleetHub) unsubscribe(ch chan struct{}) {
	h.mu.Lock()
	delete(h.subs, ch)
	h.mu.Unlock()
}

// fleetMsg is what goes over the socket. One field set per message; the client
// switches on T. Mirrors FleetMsg in web/src/lib/fleet.ts.
type fleetMsg struct {
	T        string         `json:"t"`
	Sessions []session.Meta `json:"sessions,omitempty"`
	Stats    *Stats         `json:"stats,omitempty"`
}

func (s *Server) handleFleetWS(w http.ResponseWriter, r *http.Request) {
	// Same origin policy as /ws/app: the tailnet is the auth perimeter, and the
	// hub's PWA legitimately connects to a peer's origin.
	conn, err := websocket.Accept(w, r, &websocket.AcceptOptions{OriginPatterns: []string{"*"}})
	if err != nil {
		return
	}
	defer conn.CloseNow()

	ctx := r.Context()
	ch := s.fleet.subscribe()
	defer s.fleet.unsubscribe(ch)

	stats := time.NewTicker(fleetStatsEvery)
	defer stats.Stop()

	// Read side: we expect nothing, but a reader must run or the library never
	// notices a client going away, and this goroutine would leak per closed tab.
	go func() {
		for {
			if _, _, err := conn.Read(ctx); err != nil {
				return
			}
		}
	}()

	send := func(m fleetMsg) error {
		b, err := json.Marshal(m)
		if err != nil {
			return err
		}
		wctx, cancel := context.WithTimeout(ctx, 10*time.Second)
		defer cancel()
		return conn.Write(wctx, websocket.MessageText, b)
	}

	if err := send(fleetMsg{T: "stats", Stats: s.collectStats()}); err != nil {
		return
	}

	for {
		select {
		case <-ctx.Done():
			return
		case <-ch:
			// Coalesce a burst: a turn ending moves the state, emits events and may
			// reap the session, and each of those is a change.
			timer := time.NewTimer(fleetCoalesce)
			drain := true
			for drain {
				select {
				case <-ch:
				case <-timer.C:
					drain = false
				case <-ctx.Done():
					timer.Stop()
					return
				}
			}
			if err := send(fleetMsg{T: "sessions", Sessions: s.sessionList()}); err != nil {
				return
			}
		case <-stats.C:
			if err := send(fleetMsg{T: "stats", Stats: s.collectStats()}); err != nil {
				return
			}
		}
	}
}

// sessionList is the same list GET /api/sessions returns, tagged and merged the
// same way. Shared deliberately: a push that showed a different shape from the
// fetch would be a bug nobody could see until a client mixed the two.
func (s *Server) sessionList() []session.Meta {
	metas := s.mgr.List()
	s.worktrees.tagRepos(metas)
	if s.sessionMeta != nil {
		mergeMeta(metas, s.sessionMeta.all())
	}
	return metas
}
